Issue #152 - new CollectionExtensions items: EmptyIfNull, Squash

jbatte47 opened the pull request:
- New BDD Tests
- New Patterns.Testing global bindings and models
- New methods: EmptyIfNull and Compact

jbatte47 commented:
I've updated the PR branch in response to @jlmorgan's comments. Speaking of which, if you don't mind please make your comments on the [Files Changed](https://github.com/TheTribe/code-patterns/pull/155/files) section instead of on the commit-specific page; that way the discussion won't be lost on this page.

chrislwade commented on src/Patterns/Collections/CollectionExtensions.cs:
```diff
@@ -64,5 +65,25 @@ public static void Each<TItem>(this IEnumerable<TItem> items, Action<TItem> acti
+    }
+
+    /// <summary>
+    /// Compacts the specified set by removing all null items.
```
Consider adding a note about the fact that the order of elements within the compacted set is not guaranteed to match the original set.

chrislwade commented on src/Patterns/Collections/CollectionExtensions.cs:
```diff
@@ -1,68 +1,93 @@
+    }
+
+    /// <summary>
+    ///   Compacts the specified set by removing all null items.
```
Copying my comment onto the newest commit, consider adding a note about the order of the set not being preserved.

jbatte47 replied:
Got it. I'll make the change as soon as I get home.
On Feb 7, 2014 3:26 PM, "Christopher Wade" notifications@github.com wrote:

> In src/Patterns/Collections/CollectionExtensions.cs:
> 
> > -        }
> > -      }
> > -    }
> >   +
> > -    /// <summary>
> >   
> >   
> > -    ///   Guarantees that the specified set is not null. If the set is null, an empty set is returned.
> > -    /// </summary>
> >   
> >   
> > -    /// <typeparam name="TItem">The type of the item.</typeparam>
> > -    /// <param name="set">The set.</param>
> > -    public static IEnumerable<TItem> EmptyIfNull<TItem>(this IEnumerable<TItem> set)
> > -    {
> > -      return set ?? Enumerable.Empty<TItem>();
> > -    }
> >   +
> > -    /// <summary>
> >   
> >   
> > -    ///   Compacts the specified set by removing all null items.
> 
> Copying my comment onto the newest commit, consider adding a note about
> the order of the set not being preserved.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/TheTribe/code-patterns/pull/155/files#r9555820
> .

chrislwade commented:
Looks good otherwise.

jbatte47 commented:
Actually, this is not `Where`'s fault. Upon searching for the right web resource to drop in as a link, I found that [one other method in use is known to be the culprit here](http://stackoverflow.com/questions/204505/preserving-order-with-linq), and it's one that didn't make it onto my radar because I wasn't paying attention to the evaluation scope. Turns out `Distinct` (which is called to get the first 100 _unique_ index positions from the [distribution](http://tribesoftware.org/code-patterns/blog/2014/02/06/wow-math-is-cool/) I'm using) was part of an expression that got evaluated once when printing out the contents to the SpecFlow output window, and again when `Compact` was called against it, producing two different sequences with equivalent content. I added a call to `ToArray` at the end of the `Randomize` method, thus forcing the only evaluation of `Distinct` to occur there. Now the sets show up in the same order, every time. I'll push that change up, and we won't have to adjust the comments.

jlmorgan commented:
Splendid catch!

jbatte47 committed:
Issue #152 - new CollectionExtensions items: EmptyIfNull, Squash

  - New BDD Tests
  - New Patterns.Testing global bindings and models
  - New methods: EmptyIfNull and Compact

jbatte47 commented:
I'm gonna take @chrislwade's previous consent as a go-ahead. Merging!
